File: app/src/main/java/com/example/navigationapplication/kotlin/_04kotlinInterface/_74kotlin.java
package com.example.navigationapplication.kotlin._04kotlinInterface;

/**
 * author: Angelo.Luo
 * date : 01/04/2024 3:55 PM
 * description:
 */
public class _74kotlin {
}

//懒汉式单例
class LazySingleton {
    private volatile static LazySingleton instance = null;

    private LazySingleton() {

    }

    public static LazySingleton getInstance() {
        if (instance == null) {
            synchronized (LazySingleton.class) {
                if (instance == null) {
                    instance = new LazySingleton();
                }
            }
        }
        return instance;
    }

}

//饿汉式单例
class EagerSingleton {
    private EagerSingleton() {

    }

    private final static EagerSingleton instance = new EagerSingleton();

    public static EagerSingleton getInstance() {
        return instance;
    }
}

File: app/src/main/java/com/example/navigationapplication/kotlin/_04kotlinInterface/_74kotlin.kt
package com.example.navigationapplication.kotlin._04kotlinInterface

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/**
 * author: Angelo.Luo
 * date : 01/02/2024 8:53 AM
 * description:
 */
//单例模式
fun main() {
//    SingletonClass.show()
//    SingletonClass.show()

//    println(SingletonLazyClass.getInstanceObj())
//    println(SingletonLazyClass.getInstanceObj())
//    println(SingletonLazyClass.getInstanceObj())
//    println(SingletonLazyClass.getInstanceObj())
//    println(SingletonLazyClass.getInstanceObj())
//    println(SingletonLazyClass.getInstanceObj())

    runBlocking {
        val job = launch(Dispatchers.Default) {
            repeat(100) {
                for (i in 1..100) {
                    println(SingletonLazyClass.getInstanceObj())
                }
            }
        }
        job.join()
    }


    SingletonLazyClass02.instance.show()
}

//1.饿汉式：加上object就是一个实打实的单例
object SingletonClass {
    fun show() {
        println("object is $this")
    }
}

//2.单例模式：懒汉式
class SingletonLazyClass {
    companion object {
        private var instance: SingletonLazyClass? = null
            get() {
                if (field == null) {
                    field = SingletonLazyClass()
                }
                return field
            }

        @Synchronized
        fun getInstanceObj(): SingletonLazyClass? {
            return instance!!
        }
    }
}

//2.单例模式：懒汉式
class SingletonLazyClass02 {
    companion object {
        val instance: SingletonLazyClass02 by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) { SingletonLazyClass02() }
    }

    fun show(): Unit = println("SingletonLazyClass02")
}